src/components/image-plus.js

```diff
@@ -0,0 +1,252 @@
+      // TODO: If/when THREE gets fixed, we should be able to safely remove this
+      delete texture.image;
+
+      textureCache.delete(url);
+    }
+  },
+
+  async loadGIF(url) {
+    return new Promise((resolve, reject) => {
+      // TODO: pool workers
+      const worker = new GIFWorker();
+      worker.onmessage = e => {
+        const [success, frames, delays, disposals] = e.data;
+        if (!success) {
+          reject(`error loading gif: ${e.data[1]}`);
+          return;
+        }
+
+        let loadCnt = 0;
+        for (let i = 0; i < frames.length; i++) {
+          const img = new Image();
+          img.onload = e => {
+            loadCnt++;
+            frames[i] = e.target;
+            if (loadCnt === frames.length) {
+              const texture = new GIFTexture(frames, delays, disposals);
+              texture.image.src = url;
+              resolve(texture);
+            }
+          };
+          img.src = frames[i];
+        }
+      };
+      fetch(url, { mode: "cors" })
+        .then(r => r.arrayBuffer())
+        .then(rawImageData => {
+          worker.postMessage(rawImageData, [rawImageData]);
+        })
+        .catch(reject);
+    });
+  },
+
+  loadVideo(url) {
+    return new Promise((resolve, reject) => {
+      const videoEl = createVideoEl(url);
+
+      const texture = new THREE.VideoTexture(videoEl);
+      texture.minFilter = THREE.LinearFilter;
+      videoEl.addEventListener("loadedmetadata", () => resolve(texture));
+      videoEl.onerror = reject;
+
+      // If iOS and video is HLS, do some hacks.
+      if (
+        this.el.sceneEl.isIOS &&
+        AFRAME.utils.material.isHLS(
+          videoEl.src || videoEl.getAttribute("src"),
+          videoEl.type || videoEl.getAttribute("type")
+        )
+      ) {
+        // Actually BGRA. Tell shader to correct later.
+        texture.format = THREE.RGBAFormat;
+        texture.needsCorrectionBGRA = true;
+        // Apparently needed for HLS. Tell shader to correct later.
+        texture.flipY = false;
+        texture.needsCorrectionFlipY = true;
+      }
+    });
+  },
+
+  loadImage(url) {
+    return new Promise((resolve, reject) => {
+      textureLoader.load(url, resolve, null, function(xhr) {
+        reject(`'${url}' could not be fetched (Error code: ${xhr.status}; Response: ${xhr.statusText})`);
+      });
+    });
+  },
+
+  async update() {
+    let texture;
+    try {
+      const url = this.data.src;
+      const contentType = this.data.contentType;
+      if (!url) {
+        return;
+      }
+
+      if (textureCache.has(url)) {
+        const cacheItem = textureCache.get(url);
+        texture = cacheItem.texture;
+        cacheItem.count++;
+      } else {
+        if (url === "error") {
+          texture = errorTexture;
+        } else if (contentType === "image/gif") {
+          console.log("load gif", contentType);
```

Comment:
are these console.logs necessary for production?

Reply:
no I think not